### PR TITLE
ttnn.reshape: support bfloat4_b via block-float typecast round-trip

### LINKED ISSUE #44919: ttnn.reshape rejects bfloat4_b inputs — extend support to match bfloat8_b
**Problem**

ttnn.reshape (and ops that go through it, e.g. ttnn.unsqueeze, ttnn.outer after this [PR](https://github.com/tenstorrent/tt-metal/pull/44484)) fatals when given a BFLOAT4_B input. BFLOAT8_B is supported via a typecast workaround; BFLOAT4_B is not. This prevents some ops like ttnn.outer from supporting BFLOAT4_B end-to-end and forces them to typecast manually.

**Reproduction**

```
import torch, ttnn
  device = ttnn.open_device(device_id=0)
  x = ttnn.from_torch(
      torch.rand([1, 1, 32, 32], dtype=torch.bfloat16),
      dtype=ttnn.bfloat4_b, device=device, layout=ttnn.TILE_LAYOUT,
  )
  ttnn.reshape(x, ttnn.Shape([1, 1, 32, 32, 1]))   # TT_FATAL
```

**Additional context**

  ttnn/cpp/ttnn/operations/data_movement/reshape_view/device/reshape_device_operation.cpp:25-28:

  TT_FATAL(
      input_tensor_a.dtype() == DataType::BFLOAT16 or input_tensor_a.dtype() == DataType::UINT32 or
          input_tensor_a.dtype() == DataType::FLOAT32 or input_tensor_a.dtype() == DataType::INT32,
      "Can only work with bfloat16/float32 or int32/uint32 tensors");

  BFLOAT8_B reaches this validate only because reshape.cpp pre-casts it to BFLOAT16 and post-casts the result back. See the two existing workarounds:
  - Interleaved path: ttnn/cpp/ttnn/operations/data_movement/reshape_view/reshape.cpp:384-386
  - Sharded path:    ttnn/cpp/ttnn/operations/data_movement/reshape_view/reshape.cpp:317-345

  There is no equivalent branch for BFLOAT4_B.

### PR DESCRIPTION
**Pull request by sjettTT**
## What
`ttnn.reshape` (and its callers `ttnn.unsqueeze` / `ttnn.squeeze`) raised `TT_FATAL` on
`bfloat4_b` tile tensors, while `bfloat8_b` worked. Fixes #44919.

## Root cause
The reshape device op (`reshape_view/device/reshape_device_operation.cpp`) only accepts
`BFLOAT16 / FLOAT32 / UINT32 / INT32` — the reshape kernels can't operate on packed block-float
data directly. `bfloat8_b` avoids the guard because `reshape_tiled()` wraps the op in a typecast
round-trip (`bf8 → bf16 → reshape → fill padding → bf16 → bf8`). No equivalent branch existed for
`bfloat4_b`, so it reached the guard and fatal'd.

## Solution
Generalize the four `tensor.dtype() == DataType::BFLOAT8_B` checks in `reshape_tiled()` to
`is_block_float(...)` (true for both `BFLOAT8_B` and `BFLOAT4_B`), so `bfloat4_b` takes the
identical bf16 round-trip. Add `BFLOAT4_B` to the default float pad-value branch. The device-op
dtype guard is **left unchanged** — block-float is routed *around* it, exactly as
for `bfloat8_b`. This transitively fixes `ttnn.unsqueeze` / `ttnn.squeeze` (which call
`ttnn.reshape`).

## Reasoning for this approach
Extends the established block-float handling for reshape: the bf8 round-trip originated in
**#15751**, was reworked in **#20189**, and after **#26641** attempted native bf8 reshape it was
reverted in **#27022** in favor of keeping the composite typecast round-trip. **#44115**
(repeat_interleave) uses the same `is_block_float() → bf16` pattern. This closes the same parity
gap for reshape; sibling gaps remain for permute (**#48813**) and fill_pad (**#48771**).

## Testing (Wormhole)
- Added `bfloat4_b` as a dtype axis in `test_universal_input_tm_reshape.py`, matching
  `bfloat8_b`'s scope (TILE, sharded-in, `merge_ch` / `swap_hw` / `irreg_aligned`) → 9 new cases.
- Added `test_reshape_tiled_pad_value_bfloat4_b` (interleaved) in `test_tm_reshape.py` — covers the
  block-float branch (`reshape.cpp` L497/L550) that the issue's repro actually hits; the universal
  cases only exercise the sharded branch.
- **393 passed** across `test_tm_reshape.py` + `test_universal_input_tm_reshape.py` (incl. 11 bf4
  cases: 2 interleaved + 9 universal-axis); no regression across bf16/fp32/bf8/cross-strategy.
  `base_functionality/test_reshape.py`: 323 passed / 8 skipped / 7 xfailed.
- bf4 reshape is lossless on the block data (PCC **1.0000** vs the bf4-quantized input); it measures
  ~0.993 vs the original bf16, clearing the same 0.99 tolerance as bf8.

## Files
- `ttnn/cpp/ttnn/operations/data_movement/reshape_view/reshape.cpp`
- `tests/ttnn/nightly/unit_tests/operations/data_movement/test_universal_input_tm_reshape.py`
- `tests/ttnn/nightly/unit_tests/operations/data_movement/test_tm_reshape.py`

Closes #44919

### CODE REVIEW

**Review by amorrisonTT — APPROVED** [review]

**Comment by sjettTT** [comment]
### Test validation

**CI — data-movement sanity (Wormhole `wh_n300_civ2`): ✅ passed**
https://github.com/tenstorrent/tt-metal/actions/runs/28822242064/job/85480050364

**Local Wormhole (IRD) runs with this change:**
- `test_tm_reshape.py` + `test_universal_input_tm_reshape.py` → **393 passed** (incl. 11 new `bfloat4_b` cases: 2 interleaved + 9 universal-axis)
- `base_functionality/test_reshape.py` → 323 passed / 8 skipped / 7 xfailed
- bf4 reshape is lossless vs the bf4-quantized input (PCC 1.0); ~0.993 vs the original bf16, clearing the same 0.99 bar as bf8

**Nightly reshape coverage:** these tests live under `tests/ttnn/nightly/.../data_movement`, exercised by `tt-metal-l2-nightly` (`additional_test_categories=data_movement`), which isn't auto-run on PRs. I don't have permission to dispatch it here — a maintainer can trigger it and I'll link the run.

**The 2 failing checks are unrelated flakes:**
- `t3k_fabric_tests [wh_llmbox]` — `TestT3kControlPlaneInit` (fabric didn't initialize; T3000 infra), which cascades into the route tests.
- `ttnn conv group [sim_wormhole_b0]` — `test_conv1d[…BFLOAT8_B…]` failed under parallel execution but **passed on the serial re-run**.

A re-run of those two jobs should clear them.
